insights: insights and sqlstats subsystem should consume the same types

xinhaoz opened the pull request:
Previously, insights consumed its own data type containing per-execution
statisitcs about statements and transactions for outlier analysis. These
should only be created if we determine that the execution should be cached
in our insights registry.

SQL stats and insights will now consume the same data types to process and
perform their respective recordings: `sqlstats.RecordedStmtStats` and
`sqlstats.RecordedTxnStats`, which contain per execution statistics. This
is supporting work in moving sql stats recording to to occur in the same
async routine as insights.

For reviewers: this commit mostly deals with changing logic in the insights
package and related testing to perform analysis on `sqlstats.Recorded{Stmt,Txn}Stats`
instead of `insights.{Statement,Transaction}` types. We also change the function
signatures for the sqlstats containers to accept pointers to `sqlstats.Recorded*`
objects instead of by value.

Epic: none
Part of: https://github.com/cockroachdb/cockroach/issues/141024
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141767)
<!-- Reviewable:end -->

xinhaoz commented on pkg/sql/sqlstats/insights/registry.go:
```diff
@@ -214,3 +223,110 @@ func newRegistry(
 		testingKnobs: knobs,
 	}
 }
+
```
The bodies of `makeTxnInsight` and `makeStmtInsight` are essentially what was previously in the `observeTransaction` and `observeStatement` functions in sslocal.StatsCollector, which converted `sqlstats.Recorded*` objects into the insights type.

xinhaoz committed:
insights: insights and sqlstats subsystem should consume the same types

Previously, insights consumed its own data type containing per-execution
statisitcs about statements and transactions for outlier analysis. These
should only be created if we determine that the execution should be cached
in our insights registry.

SQL stats and insights will now consume the same data types to process and
perform their respective recordings: `sqlstats.RecordedStmtStats` and
`sqlstats.RecordedTxnStats`, which contain per execution statistics. This
is supporting work in moving sql stats recording to to occur in the same
async routine as insights.

This commit mostly deals with changing logic in the insights package and related
testing to perform analysis on `sqlstats.Recorded{Stmt,Txn}Stats` instead of
`insights.{Statement,Transaction}` types. We also change the function signatures
for the sqlstats containers to accept pointers to `sqlstats.Recorded*` objects
instead of by value.

Epic: none
Part of: #141024
Release note: None

kyle-a-wong commented on pkg/sql/sqlstats/insights/registry.go:
```diff
@@ -29,10 +31,9 @@ func (r *lockingRegistry) Clear() {
 }
 
-func (r *lockingRegistry) ObserveStatement(sessionID clusterunique.ID, statement *Statement) {
-	if !r.enabled() {
```
Why don't we need this anymore?

xinhaoz replied:
Ah I think this was left in from a future commit. Put it back!

dhartunian reviewed (commented):
> Reviewed 8 of 20 files at r3, 7 of 12 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141767)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kyle-a-wong and @xinhaoz)

___
*[`pkg/sql/sqlstats/insights/registry_test.go` line 287 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141767#-OMTe5RbEiupKksgwL5w:-OMTe5RbEiupKksgwL5x:bx96pdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc4a458ff389bf12011d96db88c443ec3bb9b48f/pkg/sql/sqlstats/insights/registry_test.go#L287)):*
> ```Go
> 
> 		t.Log("expected: ", expected)
> 		t.Log("actual: ", actual)
> ```

nit: remove extra logs
___
*[`pkg/sql/sqlstats/insights/detector_test.go` line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141767#-OMTanZq3OpVUTZLVAzn:-OMTanZq3OpVUTZLVAzo:b-7evez3) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc4a458ff389bf12011d96db88c443ec3bb9b48f/pkg/sql/sqlstats/insights/detector_test.go#L132)):*
> ```Go
> 			candidateLatency time.Duration
> 			status           Statement_Status
> 			isSlow           bool
> ```

why is this `isFailed` case removed?
___
*[`pkg/sql/conn_executor_exec.go` line 4332 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141767#-OMT_NFq0C-mW9cjzd5w:-OMT_NFq0C-mW9cjzd5x:b-d6077x) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc4a458ff389bf12011d96db88c443ec3bb9b48f/pkg/sql/conn_executor_exec.go#L4332)):*
> ```Go
> 	commitLat := ex.phaseTimes.GetCommitLatency()
> 
> 	recordedTxnStats := &sqlstats.RecordedTxnStats{
> ```

I'm curious if you've looked the impact of doing this since this adds a heap allocation in the hot path. Maybe it was already heap allocated. Does the microbench show anything?


<!-- Sent from Reviewable.io -->

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141767)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @kyle-a-wong)

___
*[`pkg/sql/conn_executor_exec.go` line 4332 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141767#-OMT_NFq0C-mW9cjzd5w:-OMhc7Jg2fHK9d8ya5gA:b-gja5hq) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc4a458ff389bf12011d96db88c443ec3bb9b48f/pkg/sql/conn_executor_exec.go#L4332)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I'm curious if you've looked the impact of doing this since this adds a heap allocation in the hot path. Maybe it was already heap allocated. Does the microbench show anything?
</blockquote></details>

There was no sysbench impact for this change. We were already making a heap allocation for insights (insight.Statement and insight.Transaction objects) so this is kind of just replacing that.
___
*[`pkg/sql/sqlstats/insights/detector_test.go` line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141767#-OMTanZq3OpVUTZLVAzn:-OMhj_-U7C6Fje6PBMDJ:b-vjkuen) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc4a458ff389bf12011d96db88c443ec3bb9b48f/pkg/sql/sqlstats/insights/detector_test.go#L132)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

why is this `isFailed` case removed?
</blockquote></details>

Ah I meant to move this to just set Failed on the statement directly. Done. I removed the isFailed check since that's kind of just an assertion that the field is set in the test.
___
*[`pkg/sql/sqlstats/insights/registry_test.go` line 287 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141767#-OMTe5RbEiupKksgwL5w:-OMhjXXjEa3KasAAJz1T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc4a458ff389bf12011d96db88c443ec3bb9b48f/pkg/sql/sqlstats/insights/registry_test.go#L287)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: remove extra logs
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141767#-:-OMhuwBm1BNkiqllWAj3:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian)



<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 6 files at r5, 2 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141767#-:-OMickQPDSuFwcUjqp7X:bg7igkv)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @kyle-a-wong)



<!-- Sent from Reviewable.io -->

xinhaoz commented:
tftr!
bors r+
